sympy/series/limits.py

```diff
@@ -257,11 +257,11 @@ def doit(self, **hints):
         if e.is_Order:
             return Order(limit(e.expr, z, z0), *e.args[1:])
 
-        cdir = 0
+        cdir = S.Zero
         if str(dir) == "+":
-            cdir = 1
+            cdir = S.One
         elif str(dir) == "-":
-            cdir = -1
+            cdir = S.NegativeOne
 
```

Comment:
Are these the only possibilities?

Or is it expected that you can do e.g. `dir=I` or something?

Maybe this should not be sympified if the only possibilities are `1, 0, -1`. Instead the code that uses `cdir.is_negative` should be changed to `cdir < 0`.

Reply:
What does `cdir = 0` mean?

Reply:
> Or is it expected that you can do e.g. dir=I or something?

cdir which is "complex direction" was implemented in view of solving problems for functions that have branch cuts like sqrt, inverse trigo, logs etc, in order to specify the direction of approach in such functions, thus it was decided to have cdir as a parameter in methods like `_eval_as_leading_term` , `_eval_nseries`, but for most of the other function cdir simply sits there having not much functionality and was thus kinda given the default value of `0` ig, it is barely used , simply being passed from one routine to other as cdir=cdir,  cdir in my opinion was made in view of processing basically two functional values i.e 1 and -1 that said there are routines in some classes where cdir can have imaginary values but again those are result of internal routines. So basically though the name "complex direction" I think the only sensible values for the parameters from the POV of a user(atleast what the people who worked on this thought ) are 1 and -1, whenever they are required.
> Maybe this should not be sympified if the only possibilities are 1, 0, -1. Instead the code that uses cdir.is_negative should be changed to cdir < 0.

I too have been facing this in lot of places, for cdir as well as other places, methods like `.is_negative` were made for symbolic expressions and things go south if anyhow an interger comes to the scene, I think the approach of sympifying all user given input is the best way out and also ensuring that we do not use integers internally too.

Reply:
Okay, I see the Limit constructor requires `dir` to be either `+`, `-` or `+-`. Since `+-` was handled above here it can only be `+` or `-` so cdir will definitely be set to either -1 or 1 in this block.

So why set it to zero first here?

It would be clearer here to write:
```python
        if str(dir) == "+":
            cdir = S.One
        elif str(dir) == "-":
            cdir = S.NegativeOne
        else:
            assert False
```
Then someone reading knows that there are only two cases.

Reply:
Ahhh Oscar, Arnab sorry was caught up today . I'll address all the questions above as per my understanding tomorrow!

Reply:
Okay Answering the above doubts/questions raised (@oscarbenjamin, @arnabnandikgp)

1) What is `cdir` ?
As Arnab explains above (I think even I would have explained in a similar way) cdir stands for "complex direction"

2) Does that mean `cdir` or rather the "direction of approach" can be complex ?
Absolutely (as arnab said this would mostly come into picture for functions having branch cuts)
```
>>> asin(x + 2)._eval_nseries(x, 4, None, I)
-asin(2) + pi + sqrt(3)*I*x/3 - sqrt(3)*I*x**2/9 + sqrt(3)*I*x**3/18 + O(x**4)
>>> asin(x + 2)._eval_as_leading_term(x, None, I)
pi - asin(2)
```
These are results that were heavily tested by Kalevi and myself during my GSoC (https://github.com/sympy/sympy/pull/23715)
As I shall explain below there is also a way to justify why `asin(I*x + 2)._eval_as_leading_term(x, None, 1)` should equal to `asin(x + 2)._eval_as_leading_term(x, None, I)`

3) Numerically what values of cdir can affect a series/leadterm computation ?
As far as I know only 4 values (**1, -1, I, -I**) can affect a computation. Series and Leading term implementation for functions where cdir actually plays a role were mostly written by me and @sachin-4099 (during his gsoc in 2020) and mostly whenever we need to decide on cdir 
i) We first check if it's real or imaginary
ii) Then we check if the positive or negative (or imaginary part of cdir postive or negative if cdir is imaginary)

So technically no scalar multiplication or a shift for eg `cdir = 2`, `cdir = 2* I` or `cdir = I + 1` matters

4) So how does `cdir = 0` contribute ?
Technically it doesn't and just acts as a placeholder, cause sympy didn't give great results for series or limits for functions where actually the direction of approach matters before 2022. 

This can be shown by taking simple cases like logarithms or inverse trigs having branch cuts/points. Some tests I wrote in 2022
```
    assert log(-2*x).as_leading_term(x, cdir=1, logx=p) == p + log(2) + I*pi
    assert log(-2*x).as_leading_term(x, cdir=-1, logx=p) == p + log(2) - I*pi

    assert asin(1/x).as_leading_term(x, cdir=1) == I*log(x) + pi/2 - I*log(2)
    assert asin(1/x).as_leading_term(x, cdir=-1) == -I*log(x) - 3*pi/2 + I*log(2)
```

So I think when _eval_as_leading_term or _eval_nseries was framed, there was a vision that a parameter encapsulating "the direction of appraoch" would play a role in the future but during that time nothing was addressed hence I think a default value of `cdir = 0` was used. 

Hence according to my discussions with Kalevi (all these are well documentation on the gitter channel as we had all the discussions there) ... I can claim that `cdir = 0` implies nothing but `cdir = 1` or rather we consider "a real right hand direction of appraoch towards origin" by default just like we use dir = "+" by default for Limits.

5) When did the shift from `cdir = 0` to `cdir = 1` happen ? 

This started from the algorithm Kalevi and I framed for calculating leading term and series for simple functions like pow, sqrt, log etc that had branch cuts. These formed the fundamental building blocks for other functions have branch cuts like inverse trig, inverse hyperbolic etc.

The algorithm can be read here (https://github.com/sympy/sympy/pull/23592#issue-1262779264). Pasting an image too 

<img width="564" alt="image" src="https://github.com/user-attachments/assets/2b935ed7-4fc1-41d5-b25f-10ae9ae4dd50">

6) Lastly as I said above, we can justify as to why `asin(I*x + 2)._eval_as_leading_term(x, None, 1)` should equal to `asin(x + 2)._eval_as_leading_term(x, None, I)`

This is because as the algorithm says we first do `x = cdir*t` where where `t` is `real` and `positive` and the idea I think it to transform every direction to a "real right hand side of approach" and then do the computation.

This helps us with `dir` too as it can only be "+" or "-" . So although we don't have a complex dir like a complex cdir, we can still do 
```
>>> limit(asin(I*x + 2), x, 0, "+")
pi - asin(2)
```
and think of this being the leading term for asin(x + 2) but with a complex direction of approach.

Not sure everything I wrote above makes sense as I had worked on it a couple years back but this should be all I know. If you have any questions let me know

Reply:
Basically concluding everything I think cdir = S.One can be the new default and I don't think any test would fail if we do this !

Reply:
Let me know what y'all think is the correct way to proceed here and I would make the changes after that. I would have gone with the above comment.

Reply:
Okay then while this code here will only use `cdir = +- 1` it is also possible to have `cdir=+-I` in other places where `cdir` is used and that can happen during internal calls as well.

In that case I think that we need `cdir` to be always sympified. That can happen here but it should also be done at entry by `leadterm` and `as_leading_term`. The correct function to call for this is `_sympify` rather than `sympify` because it should not do things like string parsing etc.

Since this code will never actually allow `cdir = 0` there should not be a line that says `cdir = 0` because that is confusing. This function is a big hairy mess and should ideally be tidied up. It should have something like this at the top that clearly shows how this conditionality is being handled:
```python
dir = dir.name
if dir == `+-`:
   # one thing
elif dir == '+':
   cdir = S.One
elif dir == '-':
   cdir = S.NegativeOne
else:
   assert False
```
Other code that calls `leadterm` or `as_leading_term` can pass `cdir=0` and so we need to be clear about what this means. It is either analogous to `+-` (or rather more generally to any complex direction) or it just means `+`. Either way that needs to be clearly stated somewhere.

I think that there is some value in being able to specify *explicitly* "any complex direction" rather than getting a series that is only for a single direction like `+`. It may be that the best *default* behaviour though is to treat `cdir = 1` if that is already implicitly assumed in places...

In general I think it is unfortunate that limit defaults to a one-sided limit but that is probably too difficult to change now...

Reply:
Okay answering some questions here 

> cdir=0 and so we need to be clear about what this means. It is either analogous to +-

Nope, `cdir = 0` does not stand for `dir = "+"`. I had the same doubt during my GSoC but my mentors Kalevi and Sidharth made that clear during proposal review (hence this is documented in my proposal too)

> I think that there is some value in being able to specify explicitly "any complex direction" rather than getting a series that is only for a single direction like +. 

1) You can explicitly specify the direction of interest for series (series supports both `dir` and `cdir`). So you can have 
```
>>> asin(x + 2).series(x, n = 3, cdir = I)
-asin(2) + pi + sqrt(3)*I*x/3 - sqrt(3)*I*x**2/9 + O(x**3)
>>> asin(x + 2).series(x, n = 3, cdir = 1)
-I*log(sqrt(3) + 2) + pi/2 - I*x*(3 + 2*sqrt(3))/(3*sqrt(3) + 6) - I*x**2*(-7*sqrt(3) - 12)/(36*sqrt(3) + 63) + O(x**3)
```

All these results were verified during my GSoC so I am confident of how cdir works here.

2) I don't think we can do something like `sin(x).series(x, dir = "+", cdir = -1)`
We might not return an error if that's done but we assume the user would know that this doesn't make sense.

3) I think cdir is given more power/weightage as it is can take up more values.
`dir` can be `+`, `-` but `cdir` can be `1`, `-1`, `I`, `-I`. Hence we have 

https://github.com/sympy/sympy/blob/e676df8ded885babe62bd31bec5db8c96b7480e7/sympy/core/expr.py#L2979-L2992

So basically 
if cdir is defined, then frame dir based on cdir 
and if cdir is 0, then frame cdir based on dir.

> It may be that the best default behaviour though is to treat cdir = 1 if that is already implicitly assumed in places...

Yuss, most computations are done that way. For example series at `oo` will have a `dir = "-"` direction of approach but we transform this to series at `0` with `dir = "+"` and `cdir = 1`
https://github.com/sympy/sympy/blob/e676df8ded885babe62bd31bec5db8c96b7480e7/sympy/core/expr.py#L2996-L2999

So I can say that most calls internally take care of the transformation (substitution for `x` by `1/x` or whatever) followed by a transformation for `cdir = 1` (even for the imaginary cdir case). I think atleast for starters we should stick to this.

Reply:
> This function is a big hairy mess and should ideally be tidied up

Yes. This is surely something for the future (possibly being addressed through the grant we plan on applying).

Reply:
That being said I have some final questions for this PR (answering with my thoughts for each)

1) Should we use sympification for cdir ?
yes I think we should

2) Could we make the change `cdir = S.One` from `cdir = 0` for the default case for `_eval_as_leading_term` and `_eval_nseries` here i?
I don't see anything stopping us from doing that. The tests won't fail and also we know for sure that all fundamental calls  for series and leading term would eventually start from cdir = 1 after the transformation as I wrote above.

3) Should anything else be addressed ?
I think the above 2 would do it for this PR ( can open a separate pr taking care of cdir for nseries, series etc but can do it here too)

Reply:
> > I think that there is some value in being able to specify explicitly "any complex direction" rather than getting a series that is only for a single direction like +.
> 
> 1. You can explicitly specify the direction of interest for series (series supports both `dir` and `cdir`). So you can have

I'm not sure you understood what I meant by this. What I meant is that there should ideally be some way to say `dir/cdir = 'all'` or something and get back a single series that is valid for *all* complex directions or an error (or a Piecewise series?). This is like this:
```python
In [2]: abs(x).limit(x, 0, dir='+-')
Out[2]: 0

In [3]: sign(x).limit(x, 0, dir='+-')
...
ValueError: The limit does not exist since left hand limit = -1 and right hand limit = 1
```
except that it should be for series as well as for limits and it should be for every complex direction rather than just `+-` real directions. It could also make sense for limits but I assume that this is harder for limits than it is for series.

Most functions in SymPy can have a series around most points that is valid for all complex directions. Most other operations in SymPy are defined as being generalised to complex numbers.

Reply:
Yeah that makes sense, can be looked into in the future. Was just thinking for this PR, could you confirm the changes I propose ?
https://github.com/sympy/sympy/pull/26992#discussion_r1730700230

Reply:
> Should we use sympification for cdir ?

Yes

> Could we make the change cdir = S.One from cdir = 0 for the default case for _eval_as_leading_term and _eval_nseries here ?

The actual default value in these methods does not matter because they are never called without the arguments so making the arguments required is correct.

The question is what should the default be for `leadterm` and `as_leading_term`.

Currently on master that default is `0` and that is passed through to all methods if I understand correctly. This why I am asking what `cdir = 0` means. 

If I understand correctly changing the default here would mean a change in behaviour in some cases because `cdir = 0` is not handled as equivalent to `cdir = 1`.

Reply:
Nope I am sure cdir=0 will work equivalent to cdir=1, hence the question :)

We can change the default altogether, and again cdir being 0 doesn't technically contribute anywhere to my knowledge. I just see it as a placeholder